src/rosdiscover/observer/ros1.py

```diff
@@ -0,0 +1,135 @@
+                        for node in nodes:
+                            nodecontexts[node].provide(service,
+                                                       ros.services[service].format.fullname)
+
+                # Check if action candidates are complete (i.e., have all their topics)
+                # and add action if they are, or add the topics back in if they're not
+                self.process_action_candidates(action_server_candidates,
+                                               lambda node: nodecontexts[node].action_server,
+                                               nodecontexts)
+                self.process_action_candidates(action_client_candidates,
+                                               lambda node: nodecontexts[node].action_client,
+                                               nodecontexts)
+
+            self._nodes = nodecontexts
+            return self.summarise()
+        except TimeoutError:
+            logger.exception("failed to connect to the ROS master. Is a robot running?")
+            raise
+
+    def process_action_candidates(self,
+                                  action_candidates: Dict[str, Dict[str, ActionCandidate]],
+                                  add_function: Callable[[str], Callable[[str, str], None]],
+                                  nodecontexts: Dict[str, NodeContext]) -> None:
+        """
+        Processes action candidates, adding actions if the candidates are complete or adding
+        the topics back in if the action is incomplete.
+
+        Parameters
+        ----------
+        action_candidates: Dict[str, Dict[str, ActionCandidate]]
+            The set of candidates to process, keyed by node and action name
+        add_function: Callable[[str], Callable[[str, str], None]]
+            The function used to add the action to a node (either :code:``action_client`` or
+            :code:``action_server``
+        nodecontexts: Dict[str, NodeContext]
+            The node contexts to add actions parts to
+        """
+        for node, candidates in action_candidates.items():
+            for action_candidate in candidates.values():
+                if action_candidate.is_complete():
+                    add_function(node)(action_candidate.name, action_candidate.fmt)
+                else:
+                    logger.warning(f"Action {action_candidate.name} is incomplete")
+                    logger.debug(
+                        f"goal({action_candidate.goal}), cancel({action_candidate.cancel}), "
+                        f"status({action_candidate.status}), feedback("
+                        f"{action_candidate.feedback}), result({action_candidate.result}")
+                    action_candidate.add_unfinished_to_context(nodecontexts[node])
+
+    @staticmethod
+    def _filter_out_topics_and_nodes(
+            topics_and_nodes: Mapping[str, Collection[str]]
+    ) -> Dict[str, Collection[str]]:
```

Comment:
It would be helpful to add a very short docstring that describes the purpose of this method.